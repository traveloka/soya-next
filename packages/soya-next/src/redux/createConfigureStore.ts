import thunk from "redux-thunk";
import { applyMiddleware, combineReducers, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension/logOnlyInProduction";
import enhancer from "./storeEnhancer";
import middleware from "./middleware";

import type {
  Action,
  AnyAction,
  PreloadedState,
  ReducersMapObject,
} from "redux";
import type { SoyaNextStoreExt, SoyaNextThunkCompatStoreExt } from "../types";

export default function createConfigureStore(
  globalReducers?: ReducersMapObject<any, AnyAction>
) {
  return <
    TState = any,
    TAction extends Action<any> = AnyAction,
    TExtraArgs = any
  >(
    preloadedState?: PreloadedState<TState>,
    extraArgument?: TExtraArgs
  ) =>
    createStore<
      TState,
      TAction,
      SoyaNextStoreExt<TState, TAction> &
        SoyaNextThunkCompatStoreExt<TState, TAction, TExtraArgs>,
      any
    >(
      globalReducers
        ? combineReducers(globalReducers)
        : () => preloadedState || {},
      preloadedState,
      composeWithDevTools<any, any>(
        applyMiddleware(thunk.withExtraArgument(extraArgument), middleware),
        enhancer(globalReducers)
      )
    );
}
