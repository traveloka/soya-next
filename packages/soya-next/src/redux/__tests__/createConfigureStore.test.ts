import createConfigureStore from "../createConfigureStore";

describe("createConfigureStore", () => {
  it("should create basic configureStore", () => {
    const configureStore = createConfigureStore();
    const store = configureStore();
    store.replaceReducer({
      local: (state: any[] = []) => state,
    } as any);
    expect(store.getState()).toMatchSnapshot();
  });

  it("should create configureStore with global reducers", () => {
    const configureStore = createConfigureStore({
      global: (state: any = {}) => state,
    });
    const store = configureStore({});
    store.replaceReducer({
      local: (state: any[] = []) => state,
    } as any);
    expect(store.getState()).toMatchSnapshot();
  });
});
