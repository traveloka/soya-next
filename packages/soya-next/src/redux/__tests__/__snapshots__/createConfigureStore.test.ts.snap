// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfigureStore should create basic configureStore 1`] = `
Object {
  "local": Array [],
}
`;

exports[`createConfigureStore should create configureStore with global reducers 1`] = `
Object {
  "global": Object {},
  "local": Array [],
}
`;
